docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+### Recovery
+
+#### Current Recovery Mechanisms
+
+Gardener and upstream Kubernetes already provide recovery mechanism for node and pod recovery in case of a failure of a node. Those have been tested in the scope of a [availability zone outage simulation](#availability-zone-outage-simulation).
+
+_Machine recovery_
+
+In the seed control plane, `Kube Controller Manager` will detect that a node has not renewed its lease and after a timeout (usually 40 seconds - configurable via `--node-monitor-grace-period`) it will transition the `Node` to `Unknown` state. `Machine-Controller-Manager` will detect that an existing `Node` has transitioned to `Unknown` state and will do the following:
+* It will transition the corresponding `Machine` to `Failed` state after waiting for a duration(currently 10 mins configured via [machine-health-timeout](https://github.com/gardener/gardener-extension-provider-gcp/blob/dff3d2417ff732fcce69ba10bbe5d04e13781539/charts/internal/machine-controller-manager/seed/templates/deployment.yaml)).
+* Thereafter a `deletion timestamp` will be put on this machine indicating that the machine is now going to be terminated, transitioning the machine to `Terminating` state.
+* It attempts to drain the node first and if it is unable to drain the node, then currently after a period of 2 hours (configurable via `machine-drain-timeout`), it will attempt to force-delete the `Machine` and create a new machine.
+
+In case `Machine-Controller-Manager` is unable to delete a machine, then that machine will be stuck in `Terminating` state. It will attempt to launch a new machine and if that also fails, then the new machine will transition to `CrashLoopBackoff` and will be stuck in this state.
+
+_Pod recovery_
+
+Once `Kube Controller Manager` transitions the node to `Unknown`/`NotReady`, it also puts the following taints on the node:
+```yaml
+taints:
+- effect: NoSchedule
+  key: node.kubernetes.io/unreachable
+- effect: NoExecute
+  key: node.kubernetes.io/unreachable
+```
+
+This annotation has the following effect:
+* New pods will not be scheduled unless they have a toleration added which is all permissive or matches the effect and/or key.
+
+In case where `kubelet` cannot reach the control plane, evicting the pod is not possible without intervention. Gardener provides [garbage collection](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/gardenlet/controller/shoot/shoot_care_control.go#L343) which [forcefully deletes](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/operation/care/garbage_collection.go#L117-L138) pods after currently configured duration of 5 mins. This will force the Kube Scheduler to reschedule these pods to nodes in zones that are available.
```

Comment:
Can we safely increase this time window? Why 5 minutes today?

Reply:
I think because the `pod-eviction-timeout` is 5 minutes by default ([ref](https://kubernetes.io/docs/concepts/architecture/nodes/#condition)). Can you explain why we potentially need a higher value?